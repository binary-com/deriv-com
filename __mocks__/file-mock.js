// this file is used for mocking assets except for stylesheets
module.exports = 'test-file-stub'
