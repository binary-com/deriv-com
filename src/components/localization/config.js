import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import { str as crc32 } from 'crc-32'
import language_config from '../../../i18n-config'

i18n.use(initReactI18next).init({
    ns: ['translations'],
    defaultNS: 'translations',

    // To investigate react-i18next translation issues enable this
    debug: false,

    interpolation: {
        escapeValue: false,
    },

    react: {
        wait: true,
        useSuspense: false,
        hashTransKey(defaultValue) {
            return crc32(defaultValue)
        },
    },
})

i18n.languages = Object.keys(language_config)

export const localize = (string, values) =>
    i18n.t(crc32(string), { defaultValue: string, ...values })

export default i18n
