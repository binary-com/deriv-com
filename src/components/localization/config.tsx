/* disable-translation */
import i18n, { use, t } from 'i18next'
import { initReactI18next } from 'react-i18next'
import { str as crc32 } from 'crc-32'
import language_config from '../../../i18n-config'
import ns1 from '../../translations/en.json'

export const defaultNS = 'translations'
export const resources = {
    en: {
        ns1,
    },
} as const

use(initReactI18next).init({
    ns: ['translations'],
    defaultNS,

    // To investigate react-i18next translation issues enable this
    debug: false,

    interpolation: {
        escapeValue: false,
    },
    react: {
        useSuspense: false,
        hashTransKey(defaultValue) {
            return crc32(defaultValue)
        },
    },
    resources,
})

i18n.languages = Object.keys(language_config)

export const localize = (key: string, values?: { search: string }) => {
    let actual_key = key
    if (key?.includes('_t_')) {
        // Since I know the indices I wanna remove, substring is used instead of regex
        actual_key = key.substring(3, key.length - 3)
    }
    return t(String(crc32(actual_key)), { defaultValue: actual_key, ...values })
}

export default i18n
