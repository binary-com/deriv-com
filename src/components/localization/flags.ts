import en from 'images/svg/flags/uk.svg'
import es from 'images/svg/flags/es.svg'
import fr from 'images/svg/flags/fr.svg'
import id from 'images/svg/flags/id.svg'
import it from 'images/svg/flags/it.svg'
import pl from 'images/svg/flags/pl.svg'
import pt from 'images/svg/flags/pt.svg'
import ru from 'images/svg/flags/ru.svg'
import vi from 'images/svg/flags/vi.svg'
import th from 'images/svg/flags/th.svg'
import zh from 'images/svg/flags/zh.svg'
import tr from 'images/svg/flags/tr.svg'
import ac from 'images/svg/flags/uk.svg'
import bn from 'images/svg/flags/bn.svg'

const flags = {
    en,
    es,
    fr,
    id,
    it,
    pl,
    pt,
    ru,
    vi,
    th,
    zh,
    tr,
    ac,
    bn,
}

export default flags
