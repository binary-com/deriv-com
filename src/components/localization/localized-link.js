import React, { useContext, useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { Link as GatsbyLink } from 'gatsby'
import { AnchorLink } from 'gatsby-plugin-anchor-links'
import { LocationContext } from '../layout/location-context.js'
import language_config from '../../../i18n-config'
import { LocaleContext } from './locale-context'
import { localized_link_url } from 'common/constants'
import {
    getLocalizedUrl,
    getDerivAppLocalizedURL,
    getThaiExcludedLocale,
    replaceLocale,
} from 'common/utility'
import { DerivStore } from 'store'

export const SharedLinkStyle = css`
    color: var(--color-white);
    text-decoration: none;
    padding: 0.5rem 1rem;
    transition: text-shadow 0.25s;
    position: relative;

    &::before {
        content: '';
        position: absolute;
        transition: width 0.25s;
        height: 0.2rem;
        width: 0;
        background-color: var(--color-red);
        bottom: 0;
    }
    &:hover {
        text-shadow: 0 0 0.8px var(--color-white), 0 0 0.8px var(--color-white);

        &::before {
            width: 1.6rem;
        }
    }
    &.active {
        text-shadow: 0 0 0.8px var(--color-white), 0 0 0.8px var(--color-white);
    }

    ${(props) =>
        props.active &&
        css`
            text-shadow: 0 0 0.8px var(--color-white), 0 0 0.8px var(--color-white);

            &::before {
                width: 1.6rem;
            }
        `}
`

const ShareDisabledStyle = css`
    ${(props) =>
        props.disabled &&
        `
        pointer-events: none;
        opacity: 0.32;`}
`

const StyledAnchor = styled.a`
    ${ShareDisabledStyle}
`

const StyledAnchorLink = styled(AnchorLink)`
    ${ShareDisabledStyle}
`

const StyledGatsbyLink = styled(GatsbyLink)`
    ${ShareDisabledStyle}
`

export const LocalizedLink = React.forwardRef(({ external, ...props }, ref) => {
    const { locale } = useContext(LocaleContext)
    const [has_mounted, setMounted] = useState(false)

    useEffect(() => {
        setMounted(true)
    }, [])

    if (external || external === 'true') {
        return <ExternalLink mounted={has_mounted} locale={locale} ref={ref} {...props} />
    }

    return <InternalLink mounted={has_mounted} locale={locale} ref={ref} {...props} />
})

LocalizedLink.displayName = 'LocalizedLink'
LocalizedLink.propTypes = {
    external: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
}

const non_localized_links = ['/careers', '/academy']

const InternalLink = ({
    aria_label,
    children,
    mounted,
    has_no_end_slash,
    is_anchor,
    locale,
    to,
    ...props
}) => {
    // If it's the default language or non localized link, don't do anything
    // If it's another language, add the "path"
    // However, if the homepage/index page is linked don't add the "to"
    // Because otherwise this would add a trailing slash
    const { is_default, path } = language_config[locale]
    const is_non_localized = non_localized_links.includes(to?.replace(/\/$/, ''))
    const is_index = to === `/`
    const localized_url = getLocalizedUrl(path, is_index, to)
    const path_to = is_default || is_non_localized ? to : localized_url
    let internal_to = path_to

    // remove trailing character only if it is a forward slash
    internal_to = has_no_end_slash ? internal_to.replace(/\/$/, '') : internal_to

    if (is_anchor) {
        return (
            <StyledAnchorLink title={aria_label} to={internal_to} disabled={!mounted} {...props} />
        )
    }
    return (
        <StyledGatsbyLink aria-label={aria_label} to={internal_to} disabled={!mounted} {...props}>
            {children}
        </StyledGatsbyLink>
    )
}

InternalLink.propTypes = {
    aria_label: PropTypes.string,
    children: PropTypes.node,
    has_no_end_slash: PropTypes.bool,
    is_anchor: PropTypes.bool,
    locale: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    mounted: PropTypes.bool,
    to: PropTypes.string.isRequired,
}

const affiliate_links = ['affiliate_sign_in', 'affiliate_sign_up']
const deriv_app_links = ['dbot', 'deriv_app', 'mt5', 'derivx']
const deriv_other_products = ['binary', 'smart_trader']
const deriv_social_platforms = ['blog', 'community', 'api', 'zoho']
// add item to this array if you need to make an internal link open on a new tab without modal window
// !only for  paths without localisation: add item to this array if you need to make an internal link open on a new tab without modal window
const only_en_new_tab_no_modal = ['tnc/security-and-privacy.pdf']
const new_tab_no_modal = ['terms_and_conditions/#clients']

const getURLFormat = (type, locale, to, affiliate_lang) => {
    if (deriv_app_links.includes(type)) {
        return getDerivAppLocalizedURL(localized_link_url[type], locale, to)
    } else if (affiliate_links.includes(type)) {
        return `${localized_link_url[type]}?lang=${affiliate_lang}`
    } else if (deriv_other_products.includes(type)) {
        return `${localized_link_url[type]}/${getThaiExcludedLocale(locale)}/${to}.html`
    } else if (deriv_social_platforms.includes(type)) {
        return `${localized_link_url[type]}${to}`
    } else if (new_tab_no_modal.includes(type)) {
        return `${localized_link_url.domain_full_url}${locale === 'en' ? '' : '/' + locale}/${
            type.replace(/_/g, '-') + '/'
        }`
    } else if (only_en_new_tab_no_modal.includes(type)) {
        return `${localized_link_url.domain_full_url}/${type.replace(/_/g, '-')}`
    } else {
        return to
    }
}

const ExternalLink = ({
    aria_label,
    children,
    mounted,
    is_mail_link,
    locale,
    onClick,
    ref,
    rel,
    style,
    target,
    to,
    type,
    ...props
}) => {
    const { is_eu_country } = useContext(DerivStore)
    const { setModalPayload, toggleModal } = useContext(LocationContext)
    const { affiliate_lang } = language_config[locale]
    const url = getURLFormat(type, replaceLocale(locale), to, affiliate_lang)
    const show_modal =
        is_eu_country &&
        !is_mail_link &&
        !affiliate_links.includes(type) &&
        !deriv_app_links.includes(type) &&
        !deriv_social_platforms.includes(type) &&
        !new_tab_no_modal.includes(type) &&
        !only_en_new_tab_no_modal.includes(type)

    const default_style = { cursor: 'pointer' }
    let final_target = target
    if (new_tab_no_modal.includes(type) || only_en_new_tab_no_modal.includes(type)) {
        final_target = '__blank'
    }

    const handleClick = (e) => {
        if (show_modal) {
            e.preventDefault()
            setModalPayload({
                to: url,
                target,
                rel,
                ref,
                aria_label: aria_label,
            })
            toggleModal()
        }
        if (typeof onClick === 'function') {
            onClick(e)
        }
    }

    return (
        <StyledAnchor
            style={style ? style : default_style}
            aria-label={aria_label}
            href={!show_modal ? url : ''}
            onClick={show_modal ? handleClick : null}
            disabled={!mounted}
            target={final_target}
            rel={rel}
            {...props}
        >
            {children}
        </StyledAnchor>
    )
}

ExternalLink.propTypes = {
    aria_label: PropTypes.string,
    children: PropTypes.node,
    is_mail_link: PropTypes.bool,
    locale: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    mounted: PropTypes.bool,
    onClick: PropTypes.func,
    ref: PropTypes.string,
    rel: PropTypes.string,
    style: PropTypes.object,
    target: PropTypes.string,
    to: PropTypes.string,
    type: PropTypes.string,
}
