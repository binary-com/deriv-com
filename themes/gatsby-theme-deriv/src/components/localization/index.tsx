export * from './locale-context'
export * from './language-switcher'
export * from './localized-link'
export * from './config'
export * from './withIntl'
export * from './localize'
