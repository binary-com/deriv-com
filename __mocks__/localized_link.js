import React from 'react'
import { DerivStore } from 'store'
import { LocationProvider } from 'components/localization/locale-context'

export const MockProvider = ({ is_eu_country, children, locale }) => {
    return (
        <DerivStore.Provider
            value={{
                is_eu_country,
                is_p2p_allowed_country: true,
                crypto_config: {},
                website_status: {},
                website_status_loading: {},
                setWebsiteStatus: {},
            }}
        >
            <LocationProvider pageContext={{ locale }}>{children}</LocationProvider>
        </DerivStore.Provider>
    )
}
