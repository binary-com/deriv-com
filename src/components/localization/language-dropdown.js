import React from 'react'
import { graphql, useStaticQuery } from 'gatsby'
import PropTypes from 'prop-types'
import styled, { keyframes } from 'styled-components'
import Cookies from 'js-cookie'
import { useOutsideClick } from 'components/hooks/use-outside-click'
import { QueryImage, Text } from 'components/elements'
import { ReactComponent as Chevron } from 'images/svg/custom/chevron-bottom.svg'
import device from 'themes/device'

const Container = styled.div`
    position: relative;

    @media ${device.mobileL} {
        position: static;
    }
`

const Display = styled.div`
    display: flex;
    align-items: center;
    margin-right: 2.4rem;
    cursor: pointer;

    @media ${device.mobileL} {
        margin-right: 0.4rem;
    }
    @media ${device.mobileM} {
        margin: 0 0.2rem;
    }
`

const Arrow = styled(Chevron)`
    ${(props) => (props.expanded === 'true' ? 'transform: rotate(-180deg);' : '')}
    transition: transform 0.25s;

    & path {
        fill: var(--color-white);
    }
    @media ${device.mobileL} {
        display: none;
    }
`

const Absolute = styled.div`
    position: absolute;
    z-index: -1;
    top: ${(props) => (props.is_high_nav ? '4.8rem' : '5.5rem')};
    left: -22rem;
    height: auto;
    background-color: var(--color-white);
    transition: opacity 0.35s ease-in-out;
    cursor: default;
    border-radius: 4px;
    will-change: opacity;
    display: ${(props) => !props.is_open && 'none'};

    @media ${device.mobileL} {
        top: ${(props) => (props.is_high_nav ? '7rem' : '9rem')};
        left: 0;
    }
`

const FadeInDown = keyframes`
    from {
        opacity:0;
    }
    to {
        opacity:1;
    }
`
const FadeOutUp = keyframes`
    from {
        opacity:1;
    }
    to {
        opacity:0;
    }
`
const ItemContainer = styled.div`
    background-color: var(--color-white);
    padding: 1.6rem 0.8rem;
    position: relative;
    width: auto;
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-column-gap: 2.4rem;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.1);
    border-radius: 4px;
    will-change: opacity;
    animation-name: ${(props) => (props.is_open ? FadeInDown : FadeOutUp)};
    animation-fill-mode: both;
    animation-duration: 0.3s;
    @media ${device.mobileL} {
        width: 100vw;
        border-radius: 0;
    }

    &::after {
        content: '';
        position: absolute;
        width: 1px;
        background: var(--color-grey-8);
        height: 80%;
        top: 0;
        left: 50%;
        margin-top: 2.4rem;
        margin-bottom: 2.4rem;
    }
`

const Item = styled.div`
    display: flex;
    align-items: center;
    pointer-events: ${(props) => (props.disabled ? 'none' : 'all')};
    cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
    padding: 0.8rem 1.6rem;
    transition: background 0.25s;

    &:hover {
        background: rgba(245, 247, 250, 0.64);
    }
`

const ResponsiveText = styled(Text)`
    white-space: nowrap;

    @media ${device.mobileL} {
        display: none;
    }
`

const query = graphql`
    query {
        en: file(relativePath: { eq: "flags/en.png" }) {
            ...fadeIn
        }
        es: file(relativePath: { eq: "flags/es.png" }) {
            ...fadeIn
        }
        fr: file(relativePath: { eq: "flags/fr.png" }) {
            ...fadeIn
        }
        id: file(relativePath: { eq: "flags/id.png" }) {
            ...fadeIn
        }
        it: file(relativePath: { eq: "flags/it.png" }) {
            ...fadeIn
        }
        pl: file(relativePath: { eq: "flags/pl.png" }) {
            ...fadeIn
        }
        pt: file(relativePath: { eq: "flags/pt.png" }) {
            ...fadeIn
        }
        ru: file(relativePath: { eq: "flags/ru.png" }) {
            ...fadeIn
        }
        vi: file(relativePath: { eq: "flags/vi.png" }) {
            ...fadeIn
        }
        th: file(relativePath: { eq: "flags/th.png" }) {
            ...fadeIn
        }
        zh: file(relativePath: { eq: "flags/zh.png" }) {
            ...fadeIn
        }
        ac: file(relativePath: { eq: "flags/en.png" }) {
            ...fadeIn
        }
    }
`

const Dropdown = ({ default_option, onChange, option_list, is_high_nav }) => {
    const [is_open, setOpen] = React.useState(false)
    const dropdown_ref = React.useRef(null)

    const data = useStaticQuery(query)
    useOutsideClick(dropdown_ref, () => setOpen(false))

    const toggleVisibility = () => setOpen(!is_open)

    const closeList = () => {
        setOpen(false)
    }

    const handleSelect = (value) => {
        onChange({ target: { id: value } })
        closeList()
    }

    const default_abbreviation = default_option.path.substring(0, 2)

    return (
        <>
            <Container ref={dropdown_ref}>
                <Display onClick={toggleVisibility}>
                    <QueryImage
                        width="24px"
                        height="24px"
                        data={data[default_abbreviation]}
                        alt=""
                    />
                    <ResponsiveText color="white" ml="0.8rem" weight="bold" mr="0.4rem">
                        {default_option.short_name}
                    </ResponsiveText>
                    <Arrow expanded={`${is_open ? 'true' : 'false'}`} />
                </Display>

                <Absolute is_high_nav={is_high_nav} is_open={is_open}>
                    <ItemContainer is_open={is_open}>
                        {option_list.map((option, idx) => {
                            if (!option) return null
                            const abbreviation = option.path.substring(0, 2)
                            const current_option = default_option.path === option.path
                            return (
                                <Item
                                    disabled={current_option}
                                    id={option.value}
                                    onClick={() => {
                                        handleSelect(option.value)
                                        Cookies.set('lang_is_fixed', 'true')
                                    }}
                                    key={idx}
                                >
                                    <QueryImage
                                        width="24px"
                                        height="24px"
                                        data={data[abbreviation]}
                                        alt=""
                                    />
                                    <Text ml="0.8rem" color={current_option ? 'red' : 'black'}>
                                        {option.text}
                                    </Text>
                                </Item>
                            )
                        })}
                    </ItemContainer>
                </Absolute>
            </Container>
        </>
    )
}

Dropdown.propTypes = {
    default_option: PropTypes.object,
    is_high_nav: PropTypes.bool,
    onChange: PropTypes.func,
    option_list: PropTypes.array,
}

export default Dropdown
