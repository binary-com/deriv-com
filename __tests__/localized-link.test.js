import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { LocalizedLink } from '../src/components/localization/localized-link'
import { MockProvider } from '../__mocks__/localized_link'

describe('Localized Link Tests', () => {
    describe('External Localized Links', () => {
        it('renders correctly when flag is passed', () => {
            const rendered_link = renderer
                .create(
                    <MockProvider>
                        <LocalizedLink external to="/">
                            Test Link
                        </LocalizedLink>
                    </MockProvider>,
                )
                .toJSON()
            expect(rendered_link).toMatchSnapshot()
        })
        describe('EU Countries', () => {
            describe('Modal Tests', () => {
                it('modal is shown for eu country when links are clicked for supported links', () => {
                    render(
                        <MockProvider>
                            <LocalizedLink external to="/">
                                Link
                            </LocalizedLink>
                        </MockProvider>,
                    )
                    userEvent.click(screen.getByText('Link'))
                })
                it('modal is not shown for mail links', () => {})
                it('modal is not shown for deriv app links', () => {})
                it('modal is not shown for binary links', () => {})
                it('modal is not shown for social links', () => {})
                it('modal is not shown for terms and conditions links', () => {})
                it('modals redirects to the correct link', () => {})
            })
            describe('Localization Tests', () => {
                it('correct deriv app localized link', () => {})
                it('correct binary localized link', () => {})
                it('correct affiliate localized link', () => {})
                it('correct terms and conditions localized link', () => {})
                it('correct social localized link', () => {})
            })
            it('terms and conditions link does not open in a new tab', () => {})
        })
        describe('Non EU Countries', () => {
            it('modal is not shown for non eu countries', () => {})
        })
    })
    describe('Internal Localized Links', () => {
        it('renders correctly when flag is passed', () => {
            const rendered_link = renderer
                .create(
                    <MockProvider>
                        <LocalizedLink internal to="/">
                            Test Link
                        </LocalizedLink>
                    </MockProvider>,
                )
                .toJSON()
            expect(rendered_link).toMatchSnapshot()
        })
    })
})
