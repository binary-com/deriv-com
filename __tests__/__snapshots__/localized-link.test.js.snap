// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Localized Link Tests External Localized Links renders correctly when flag is passed 1`] = `
<a
  className="sc-bdfBQB imoVvB"
  disabled={true}
  href="/"
  onClick={null}
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  Test Link
</a>
`;

exports[`Localized Link Tests Internal Localized Links renders correctly when flag is passed 1`] = `
<a
  aria-current="page"
  className="sc-dlfnuX cvSRZk"
  disabled={false}
  href="/"
  internal={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  style={Object {}}
>
  Test Link
</a>
`;
